Adding support for Query parameters in url.

ralpha opened the pull request:
Fixes #25
Adds query parameters in url. All parameters are required by default and 
using `Option<..>` will make the parameter optional. 
This is also how rocket expects the values.
As noted this fixes issue #25 where this was requested.
Also add this to the example.

ralpha commented:
Fixed/static query parameters are not parsed and this will break documentations "Try it out". But this is the same now. So it is not really an error here, but just noting it here in case.
This also does not change the `path` in the documentation. This is the same as in the tested swagger examples/demos.

ralpha commented:
Found some thing that are wrongly implemented. (it works but with more advanced use found some problems, will fix in the next few days)

ThouCheese commented:
Hi @ralpha, thank you so much for you PR! I have some problems running the code however. I tried the following:
```rust
#![feature(decl_macro, proc_macro_hygiene)]

use rocket_okapi::{openapi, swagger_ui::{make_swagger_ui, SwaggerUIConfig}};
use rocket::get;

#[openapi]
#[get("/?<id>")]
fn get_by_id(id: i32) -> String {
    format!("got {}", id)
}

fn get_docs() -> SwaggerUIConfig {
    use rocket_okapi::swagger_ui::UrlObject;

    SwaggerUIConfig {
        urls: Some(vec![
            UrlObject::new("Main", "/openapi.json"),
        ]),
        ..Default::default()
    }
}

fn main() {
    rocket::Rocket::ignite()
        .mount("/", rocket_okapi::routes_with_openapi![get_by_id])
        .mount("/swagger", make_swagger_ui(&get_docs()))
        .launch();
}
```
where I depend on your code like so:
```toml
[dependencies]
rocket = "0.4"
rocket_okapi = { git = "https://github.com/ralpha/okapi", default-features = false }
okapi = { git = "https://github.com/ralpha/okapi", default-features = false, features = ["derive_json_schema"] }
```
When I go to http://localhost:8000/swagger or http://localhost:8000/openapi.json however, I see no information about the query parameter I've added. Am I doing something wrong?

ThouCheese commented:
Hi @ralpha, thank you so much for your PR! It looks very good, and it seems to work pretty well! I'll go test it a bit more this evening. Would it be complicated to add support for patterns like so:
```rust
#[derive(FromForm, JsonSchema)]
struct Data {
    /// The id
    id: i32
}

/// Some endpoint
#[openapi]
#[get("/?<query..>")]
fn get_by_id(query: Form<Data>) -> String {
    format!("got {}", query.id)
}
```
This way the description field of the parameter could also be populated. I feel like this would be significantly harder to implement though, but I could be wrong.

ralpha commented:
This is how I found the [problems mentioned above](https://github.com/GREsau/okapi/pull/26#issuecomment-630541255). I was implementing my own [Query Guard](https://api.rocket.rs/v0.4/rocket/request/trait.FromQuery.html) (similar to the `Form<...>` in your case). The thing is that these structs can have multiple fields. And thus multiple query parameters have to be added. This is currently not supported in the code so requires some other changes.
I can probably get the possible fields from the JsonSchema trait.

I'm working on this but might still take some time. It should be done in not to long, but will see what happens. I'll let you know when I got something working.

ralpha commented:
Added support for Query Guards. Ex: `/?<user..>`

This allows multiple query parameters to be added to the list of
parameters in the documentation.
It also allows custom Query Guards to be created.
Code for this is included in documentation.
It also functions as a test.

There might be a problem with this code if more structs use the
same name in different modules and both implement `JsonSchema`.
The reference system might not work in that case.

Query options are required unless it is inside `Option<..>`. It also checks Optional elements in the Query Guard and marks them as Optional without marking the others.

Known issue: Query parameters with `#[serde(rename_all = "camelCase")]` break the filling of the object. The option will be `userId` in the documentation and will return it using `/?userId=1` but should be `/?user_id=1`.

@ThouCheese :smiley:

ralpha commented:
Ready to be merged again. (if I find problems I'll post them here and fix code.)

ralpha commented:
The rename issue might be able to be resolved with this:
https://rocket.rs/v0.4/guide/requests/#field-renaming
That is by the user, not this crate.
```rust
#[derive(FromForm, JsonSchema)]
#[serde(rename_all = "camelCase")]
struct External {
    #[form(field = "userId")]
    user_id: String
}
```

GREsau commented on rocket-okapi/src/request/from_query_multi_param_impls.rs:
```diff
@@ -0,0 +1,153 @@
+        match schema {
+            Schema::Object(schema) => {
+                if let Some(object) = schema.object{
+                    let object: &'r mut ObjectValidation = Box::leak(object);
+                    properties = object.properties.clone();
```
I don't understand why we're leaking the box here? Also, since the schema is owned rather than borrowed, there's no need to clone the properties - we can just take them:
```suggestion
                    properties = object.properties;
```

ralpha replied:
I'm not very familiar with `Box` yet. But this code was no longer needed after changing to `json_schema_no_ref`. So that resolves this too.

ralpha commented on rocket-okapi/src/gen.rs:
```diff
@@ -55,6 +55,20 @@ impl OpenApiGenerator {
+        }
+        None
+    }
+
```
Did not know this existed. Thanks, much nicer now.

ralpha commented on rocket-okapi/src/request/from_query_multi_param_impls.rs:
```diff
@@ -0,0 +1,153 @@
+    }
+    if let Some(schema) = referenced_schema{
+        // Get a list of properties from the structure.
+        let mut properties: BTreeMap<String, Schema> = BTreeMap::new();
```
Done

ralpha commented:
Resolved all the suggestions. Much nicer now. But rustfmt made `OpenApiFromQuery` look a bit weird. But everything else is formatted too. 
Also rebased to make sure there are no conflicts and fixed one of the tests.

ThouCheese commented on rocket-okapi/src/request/from_query_multi_param_impls.rs:
```diff
@@ -0,0 +1,131 @@
+        parameter_list.push(Parameter {
+            name: key,
+            location: "query".to_owned(),
+            description: None,
```
I suggest changing this to `description: prop_schema.metadata.as_ref().and_then(|m| m.description.clone()),`, so as to also include the description.

ralpha replied:
Oh did not even notice is was not printing the schema description.
Great suggestion!

GREsau commented:
Approved, merged, and now published to crates.io in rocket_okapi 0.5.1

Thanks so much for this @ralpha!

ralpha commented:
Thanks for merging :)
I still have some custom code in rocket_okapi so can not switch to the version of crates.io yet.
(and currently for developing and testing new features I also test it in my application that I'm making)
But as soon as it has some other features implemented I'll switch over to crate.io version.
(just switched okapi over to crate.io version :) )
